Add documentation for MSTEST0007 analyzer rule (#39386)
dotnet/docs: docs/core/testing/mstest-analyzers/mstest0007.md

---

> title: "MSTEST0007: Ensure test attributes can only be set on methods marked with TestMethod attribute"

Reviewer: We usually put the rule title so it would be

```suggestion
title: "MSTEST0007: [<TestAttribute>] can only be set on methods marked with [TestMethod]"
```

description: "Learn about code analysis rule MSTEST0007: Ensure test attributes can only be set on methods marked with TestMethod attribute"
ms.date: 02/01/2024
f1_keywords:
- MSTEST0007
- UseAttributeOnTestMethodAnalyzer
helpviewer_keywords:
- UseAttributeOnTestMethodAnalyzer
- MSTEST0007
author: cvpoienaru
ms.author: codrinpoienaru
dev_langs:
- CSharp
- VB
---

> # MSTEST0007: Ensure test attributes can only be set on methods marked with TestMethod attribute

Reviewer: ```suggestion
# MSTEST0007: [<TestAttribute>] can only be set on methods marked with [TestMethod]
```

| Property                            | Value                                                                              |
|-------------------------------------|------------------------------------------------------------------------------------|
| **Rule ID**                         | MSTEST0007                                                                         |

> | **Title**                           | Ensure test attributes can only be set on methods marked with TestMethod attribute |

Reviewer: ```suggestion
| **Title**                           | [<TestAttribute>] can only be set on methods marked with [TestMethod] |
```

| **Category**                        | Usage                                                                              |
| **Fix is breaking or non-breaking** | Non-breaking                                                                       |
| **Enabled by default**              | Yes                                                                                |
| **Default severity**                | Info                                                                               |

> | **Introduced in version**           | 3.3.0 ??                                                                           |

Author: @Evangelink is this the correct version of testfx that will contain the analyzer ?

Author: Confirmed version is 3.3.0.

## Cause

A method not marked with `[TestMethod]` attribute has one or more test attributes applied to it.

## Rule description

> Test attributes such as `[Owner]`, `[Priority]`, `[TestProperty]`, `[WorkItem]`, `[Description]`, `[ExpectedException]`, `[CssIteration]` and `[CssProjectStructure]` should only be applied on methods marked with `[TestMethod]` attribute.

Reviewer: I would do bullet points for the list of attributes (and order them alphabetically for convenience):

```suggestion
Test attributes:
- `[CssIteration]`
- `[CssProjectStructure]` 
- `[Description]`
- `[ExpectedException]`
- `[Owner]`
- `[Priority]`
- `[TestProperty]`
- `[WorkItem]`

should only be applied on methods marked with `[TestMethod]` attribute.
```

---

> title: "MSTEST0007: [<TestAttribute>] can only be set on methods marked with [TestMethod]"

Reviewer: Just wondering if we should have `Test attributes .... on test methods`

Author: This is the way it was originally (more or less, it was `Ensure test attributes ...`) but then there was a suggestion from you to change it to `[<TestAttribute>] ...`. Would you like to change it back ?

Author: Updated as per our offline sync.

---
title: "MSTEST0007: [<TestAttribute>] can only be set on methods marked with [TestMethod]"

> description: "Learn about code analysis rule MSTEST0007: Ensure test attributes can only be set on methods marked with TestMethod attribute"

Reviewer 2: ```suggestion
description: "Learn about code analysis rule MSTEST0007: Ensure test attributes are set only on methods marked with the `TestMethod` attribute"
```

ms.date: 02/01/2024
f1_keywords:
- MSTEST0007
- UseAttributeOnTestMethodAnalyzer
helpviewer_keywords:
- UseAttributeOnTestMethodAnalyzer
- MSTEST0007
author: cvpoienaru
ms.author: codrinpoienaru

> dev_langs:

Reviewer 2: Lines 13-15 aren't needed since there are no code examples.

Author: Removed, thank you !

| Property                            | Value                                                                              |
|-------------------------------------|------------------------------------------------------------------------------------|
| **Rule ID**                         | MSTEST0007                                                                         |
| **Title**                           | `[<TestAttribute>]` can only be set on methods marked with `[TestMethod]`          |
| **Category**                        | Usage                                                                              |
| **Fix is breaking or non-breaking** | Non-breaking                                                                       |
| **Enabled by default**              | Yes                                                                                |
| **Default severity**                | Info                                                                               |
| **Introduced in version**           | 3.3.0                                                                              |

## Cause

> A method not marked with `[TestMethod]` attribute has one or more test attributes applied to it.

Reviewer 2: ```suggestion
A method that's not marked with <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestMethodAttribute> has one or more test attributes applied to it.
```

## Rule description

> Test attributes:

Reviewer 2: ```suggestion
The following test attributes should only be applied on methods marked with the <xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestMethodAttribute> attribute:
```

- `[CssIteration]`
- `[CssProjectStructure]`
- `[Description]`
- `[ExpectedException]`
- `[Owner]`
- `[Priority]`
- `[TestProperty]`
- `[WorkItem]`

> should only be applied on methods marked with `[TestMethod]` attribute.

Reviewer 2: ```suggestion
```

> 

Reviewer 2: ```suggestion
```

## How to fix violations

> To fix a violation of this rule, either convert the method on which you applied one or more test attributes to a test method by setting the `[TestMethod]` attribute, or remove the test attribute(s) altogether.

Reviewer 2: ```suggestion
To fix a violation of this rule, either convert the method on which you applied the test attributes to a test method by setting the `[TestMethod]` attribute or remove the test attributes altogether.
```

## When to suppress warnings

> Do not suppress a warning from this rule. Ignoring this rule will result in your attributes being ignored since they were designed only for use in a test context.

Reviewer 2: ```suggestion
Do not suppress a warning from this rule. If you ignore this rule, your attributes will be ignored since they are designed for use only in a test context.
```
